common/crypto.hpp

```diff
@@ -15,6 +15,8 @@
 #ifndef MENDER_COMMON_CRYPTO_HPP
 #define MENDER_COMMON_CRYPTO_HPP
 
+#include <common/crypto/platform/openssl/openssl_config.h>
```

Comment:
~This cannot be right: the platform agnostic header shouldn't know about openssl.~

Edit: too bad that we need this include for the definitions below. At least it should be guarded with `ifdef MENDER_CRYPTO_OPENSSL`

Reply:
Yeh, issue is the `PrivateKey` needs to keep the `HSM` resources alive, so the `ENGINE` or the `PROVIDER`.

I have been thinking about how to abstract this away, but I can't think of any good ways atm.

All open to suggestions here ofc

Reply:
One thing which occurred to me after our talk: Do the types need to be qualified? They are just pointers, so can't they be forward declared with `class ENGINE;` and that should be enough?

Reply:
I was pondering that all evening last night!

I'm not sure if it's a good idea though, is it considered kosher?

Reply:
Yes, this is quite common, actually. It doesn't apply to us, but in public library development there is an important function to doing so. If your private struct is exposed in the header, there is a chance it could, maybe unintentionally, be used directly in the source files of whoever is using your library. This is bad, because it binds the memory layout of that struct to the binary that was built, instead of to the library. This means you can no longer update the struct, because you risk breaking compatibility with already released binaries. OTOH, if you only forward declare it, then any such unintentional use would be flagged by the compiler as illegal, and the users would be forced to use the factory functions you provide to get a pointer to the struct.